### PR TITLE
Introduce `HashCompressor` for name compression

### PR DESCRIPTION
**Pull request by bal-e**
I noticed that 'TreeCompressor' will likely have a lot of overhead, as it has to copy each label into its internal tree and it maintains a separate hashmap for every known name (where each name likely has one or two parents at the most, even though hashmaps have a large capacity).

'HashCompressor' is an alternative name compressor built on top of the 'hashbrown' crate, which offers lower-level hash table access that lets entries indirectly reference the built message for hashing and equality checks.  It maintains a single hashmap, it should be faster, and it does not require as much memory.

The implementation here makes it clear that domain names should have been stored in reverse order, at least in the wire format.  A decent chunk of the logic goes into correctly reversing the domain name.  In fact, the implementation likely quadratic runtime because of this (at least when 'Name' or 'RelativeName' are used, as backward iteration on them has quadratic runtime).

While 'std' essentially exposes the same data structures as 'hashbrown', it does not currently offer a way to perform low-level hash table access (although there is a nightly "raw entry" feature).  This may change in the future.

If not for the added 'hashbrown' dependency, I would suggest deprecating 'TreeCompressor' entirely.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -46,6 +47,7 @@ tracing-subscriber = { version = "0.3.18", optional = true, features = ["env-fil
 [features]
 default     = ["std", "rand"]
 bytes       = ["dep:bytes", "octseq/bytes"]
+hashbrown   = ["dep:hashbrown", "std"]
```
[the remark this answers is not in the document]
Comment (bal-e) [reply Cargo.toml:50]: Maybe I'll add a separate `hash-compressor` feature that enables `hashbrown`?  I suspect low-level hash maps could be useful in some other contexts (or potentially outside `std`).

File: src/base/message_builder.rs [hunk 2]
```diff
@@ -2354,6 +2360,247 @@ impl<Target: Composer> Truncate for TreeCompressor<Target> {
...
+/// [`StreamTarget`]: struct.StreamTarget.html
+#[cfg(feature = "hashbrown")]
+#[derive(Clone, Debug)]
+pub struct HashCompressor<Target> {
+    /// The underlying octetsbuilder.
+    target: Target,
+
+    /// The names inserted into the message.
+    ///
+    /// Each entry represents a name with at least one non-root label and its
```
[the remark this answers is not in the document]
Comment (bal-e) [reply src/base/message_builder.rs:2398]: Yeah, I hadn't noticed that it's unfinished!  I've rewritten it now.

File: Cargo.toml (hunk 1 above)
[the remark this answers is not in the document]
Comment (partim) [reply Cargo.toml]: I actually agree with your proposal to replace `TreeCompressor` with the new compressor when we do the next breaking release. I don’t mind the extra dependency (we could keep it optional and pull it in as part of the `std` feature).

Which is to say: maybe only require the `std` feature and leave it a that?

[the remark this answers is not in the document]
Comment (bal-e) [reply Cargo.toml]: Yeah, can do!

[the remark this answers is not in the document]
Comment (bal-e) [reply Cargo.toml]: Yep, done!

**Comment by ximon18** [comment]
I'll take a look as soon as I can.

**Comment by ximon18** [comment]
> @ximon18: Do you think this is a breaking change? It doesn't remove any existing APIs, we just discussed that `TreeCompressor` could be replaced by this for the next breaking-change release. I'd like to get it merged soon.

It seems to be purely additive so I don't see why it would be a breaking change. The only observation I had when scanning it is that I saw an `assert!()` call which is not documented in the RustDoc as a known way the code can panic... and ideally it wouldn't panic on user input, especially for a function that is able to error out. I can't tell if `hashbrown` as an added dependency in some way limits the build targets compared to now, but given that hashbrown is used by the Rust standard library I assume it is fine. I don't know how you chose the feature flags to use for hashbrown, I don't see them documented in the crate docs.

**Comment by bal-e** [comment]
@ximon18 the `assert` can only fail if somebody implements `ToName` incorrectly, which is a programming error and not something external users can cause.  The `hashbrown` feature flags are there to disable some unnecessary default features to speed up compilation.

**Pull request merged**